File: exercises/exercise-1-testing-java/src/main/java/de/emaarco/example/application/service/newsletter/SendWelcomeMailService.java
package de.emaarco.example.application.service.newsletter;

import de.emaarco.example.application.port.inbound.newsletter.SendWelcomeMailUseCase;
import de.emaarco.example.application.port.outbound.newsletter.NewsletterSubscriptionRepository;
import de.emaarco.example.domain.SubscriptionId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class SendWelcomeMailService implements SendWelcomeMailUseCase {

    private static final Logger log = LoggerFactory.getLogger(SendWelcomeMailService.class);

    private final NewsletterSubscriptionRepository repository;

    public SendWelcomeMailService(NewsletterSubscriptionRepository repository) {
        this.repository = repository;
    }

    @Override
    public void sendWelcomeMail(SubscriptionId subscriptionId) {
        var subscription = repository.find(subscriptionId);
        log.info("Sending welcome mail to {}", subscription.email());
    }

}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/de/emaarco/example/application/service/newsletter/SendWelcomeMailService.kt
package de.emaarco.example.application.service.newsletter

import de.emaarco.example.application.port.inbound.newsletter.SendWelcomeMailUseCase
import de.emaarco.example.application.port.outbound.newsletter.NewsletterSubscriptionRepository
import de.emaarco.example.domain.SubscriptionId
import mu.KotlinLogging
import org.springframework.stereotype.Service

@Service
class SendWelcomeMailService(
    private val repository: NewsletterSubscriptionRepository,
) : SendWelcomeMailUseCase {

    private val log = KotlinLogging.logger {}

    override fun sendWelcomeMail(subscriptionId: SubscriptionId) {
        val subscription = repository.find(subscriptionId)
        log.info { "Sending welcome mail to ${subscription.email}" }
    }
}